### PR TITLE
Simplify the filters by using nested iterators

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

The trait design is a bit overcomplicated, so this simplifies it by using nested iterators.

### CODE REVIEW

File: crates/iota-genesis-builder/src/stardust/process_outputs.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ use iota_sdk::types::{
     },
 };
 use iota_types::timelock::timelock::is_vested_reward;
-use tracing::info;
+use tracing::debug;
```
Comment (miker83z) [inline crates/iota-genesis-builder/src/stardust/process_outputs.rs:20]: Nice! I didn't think about it

File: crates/iota-genesis-builder/src/stardust/process_outputs.rs [hunk 2]
```diff
@@ -78,110 +66,69 @@ struct OutputHeaderWithBalance {
     balance: u64,
 }
 
-// Define the trait for filtering logic
-trait HornetFilter<Item> {
```
Comment (miker83z) [inline crates/iota-genesis-builder/src/stardust/process_outputs.rs:82]: Ok the trait itself is not needed. However...

File: crates/iota-genesis-builder/src/stardust/process_outputs.rs [hunk 3]
```diff
@@ -78,110 +66,69 @@ struct OutputHeaderWithBalance {
...
-struct HornetFilterIterator<I, Item>
-where
-    I: Iterator<Item = Item>,
-{
-    /// The vector of filters.
-    filters: Vec<Box<dyn HornetFilter<Item>>>,
+struct HornetFilterIterator<I> {
     /// Iterator over `(OutputHeader, Output)` pairs.
     outputs: I,
+    /// Stores aggregated balances for eligible addresses.
```
Comment (miker83z) [inline crates/iota-genesis-builder/src/stardust/process_outputs.rs]: ... I think in some way we need to make it simple to separate the scope of each filter. The original idea was to have a quick way to add or remove filters:
- e.g., imagine we need a new filter in a couple of sprints, with the original idea I only needed to create a new type implementing the trait.
- at the same tipe I can test out locally the different filters by simply modifying the `vec![scale_amount_filter, vesting_filter, voting_filter]` in `process_outputs_for_iota` or create dedicated `test_process_outputs` functions.

File: crates/iota-genesis-builder/src/stardust/process_outputs.rs [hunk 4]
```diff
@@ -247,97 +198,66 @@ impl HornetFilter<Result<(OutputHeader, Output)>> for ScaleIotaAmountFilter {
...
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(output) = self.filter_output() {
+            return Some(output);
+        }
+
+        // Now that we are out of the loop we collect the processed outputs from the
+        // filters
+        let next = self.unlocked_address_balances.pop_first().map(
```
Comment (miker83z) [inline crates/iota-genesis-builder/src/stardust/process_outputs.rs]: In here part of the logic of 1 filter is moved into the `next()` method. Ideally the filter should contain all its logic inside his process/filter method.

**Review by miker83z — COMMENTED** [review]
I think it is ok to remove the trait and implement your optimizations. However, the original intent was to keep the filters modular, allowing them to be easily added or removed during iteration—for instance, when adding new filters in the future or disabling some for testing. Could you pls explore a middle ground that keeps this flexibility while incorporating your changes? 🙏🏻

**Comment by DaughterOfMars** [comment]
> the original intent was to keep the filters modular, allowing them to be easily added or removed during iteration

I guess I just don't see the value in this given the added complexity. It is over-designed, IMO, for a situation that may never happen anyway and honestly wouldn't be difficult to add on top of this simpler design anyway. I don't think there is a middle-ground which would satisfy either simplicity or modularity well.

**Review by miker83z — APPROVED** [review]
LGTM!

**Pull request merged**